doc/man/man8/m2uaasd.8.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man8/m2uaasd.8.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH M2UAASD 8 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Administration"
.SH NAME
.B m2uaasd
\- M2UA-AS Configuration Daemon
.\"
.\"
.SH SYNOPSIS
.PP
.HP 12
\fBm2uaasd\fR [\fIoptions\fR]
.PD 0
.HP
\fBm2uaasd\fR {\fB-h\fR|\fB--help\fR}
.HP
\fBm2uaasd\fR {\fB-V\fR|\fB--version\fR}
.HP
\fBm2uaasd\fR {\fB-C\fR|\fB--copying\fR}
.PD
.\"
.\"
.SH DESCRIPTION
.PP
.B m2uaasd
is a configuration daemon for the M2UA-AS multiplexing driver,
.BR m2ua_as (4).
Its purpose is to open a configuration file, establish SCTP assocations between the local host and
remote SG hosts, link the SCTP associations under the M2UA-AS multiplexing driver at the appropriate
ASP positions, configure each ASP using input-output controls and then wait in the background
listening for management events on a management stream.  Also, the daemon watches for updates to the
configuration file and applies the updates as they occur.
.PP
The
.B m2uaasd
is intended to be run at system startup from a System V init script.  Once the configuration of the
M2UA-AS multiplexing driver is complete, applications programs can open, attach, enable and use
Signalling Link streams opened on each of the configured ASPs.
.\"
.\"
.SH OPTIONS
.PP
.TP
.BR -h ", " --help
When this option is encountered,
print
usage
information to
.I stdout
and exit.
.TP
.BR -V ", " --version
When this option is encountered,
print
version
information to
.I stdout
and exit.
.TP
.BR -C ", " --copying
When this option is encountered,
print
copyright and copying
information to
.I stdout
and exit.
.\"
.\"
.SH "FILE FORMAT"
.PP
When the
.BR -f " or " --file
.I filename
options is specified, the supplied file contains M2UA configuration entries as follows:
.IP \(bu \w'\(bu\(em'u
Each line in the file contains one entry or comment per line.
.IP \(bu
Each entry consists of a number of fields delimited by whitespace (see
.RB isspace (3)).
.IP \(bu
Any field that begins with a
.RI ` # '
character is a comment that continues to the end of the line.
.IP \(bu
Any line containing only a comment (beginning with optional whitespace and a
.RI ` # '
character) is ignored.
.PP
Each line containing an M2UA configuration specification entry consists of the following fields:
.TP
(1) \fIcommand\fR
.PP
A Signalling Gateway entry is formatted as follows:
.RS
.TP
(1) \fBsg\fR
A command indicating that this is an SG entry.
.TP
(2) \fIindex\fR
A unique index number assigned to the SG.
.TP
(3) \fIclei\fR
A common language location identifier or common language equipment identifier for the SG.
.TP
(4) \fIoptions\fR
A comma separate option string indicating the options associated with the SG.  Options include
whether dynamic or static configuration is required or supported.
.TP
(5) \fIaddress\fR
The (SCTP) transport address at which this SG can be reached.  This address can be a hostname or a
numbers and dots IP address and port number.
.RE
.PP
An Application Service Process entry is formatted as follows:
.RS
.TP
(1) \fBasp\fR
A command indicating that this is an ASP entry.
.TP
(2) \fIindex\fR
A unique index number assigned to the ASP.  This index number is the minor device number on the
M2UA-AS driver that will be opened to obtain an AS Stream associated with this ASP.
.TP
(3) \fIclei\fR
A common language location identifier or common language equipment identifier for the ASP.
.TP
(4) \fIaspid\fR
An ASP identifier.  When no ASP identifier is used (the SG can determine the ASP Id from the
trasnport address with which the ASP is connecting), this field will contain a single dash,
.RI ` - '.
.TP
(5) \fIsg\fR
An SG index or SG CLEI identifying the SG serving this ASP.
.TP
(6) \fIaddress\fR
The (SCTP) transport address to which this ASP will bind.  This address can be a hostname or a
numbers and dots IP address and port number.
.TP
(7) \fIstreams\fR
The number of outbound streams that the ASP will request when it connects.
.RE
.PP
A Signalling Link profile entry is formatted as follows:
.RS
.TP
(1) \fBsl\fR
A command indicating that this is an SL profile entry.
.TP
(2) \fIprofile\fR
A profile name uniquely identifying this profile and distinguishing it from all others.
.TP
(3) \fIpvar\fR
The protocol variant.
.TP
(4) \fIpopt\fR
Protocol options.
.TP
(5) \fIsif-min\fR
The minimum size of the service information field in the SS7 SDU.
.TP
(6) \fIsif-max\fR
The maximum size of the service information field in the SS7 SDU.
.RE
.PP
An Application Server entry is formatted as follows:
.RS
.TP
(1) \fBas\fR
A command indicating that this is an AS entry.
.TP
(2) \fIasp\fR
An ASP index or ASP CLEI identifying the ASP serving this AS.
.TP
(3) \fIppa\fR
The physical point of appearance identifier for the AS.  This is the PPA address to which an SL
stream will attach to attach to this application server.  This field is any 32-bit integer number.
When dynamic registration is used, the high-order 16-bits corresponds to the SDTI and the low-order
16-bits corresponds to the SDLI used in the link key for registration.  When static registration is
used, this field can contain a single dash,
.RI ` - ',
to indicate that the IID is used as the PPA.
.TP
(4) \fIiid\fR
The integer or text interface identifier for the the AS.  When the serving SG supports dynamic
configuration, and this field is set to a single dash,
.RI ` - ',
dynamic registration will be used to determine the IID for the AS.  The SDLI and SDTI that are used
to register the AS will be derived from the high- and low-order 16 bits of the PPA.
.TP
(5) \fItmode\fR
The traffic mode for the application server.  Values can be:
.BR override ", " loadshare " or " broadcast .
This is the traffic mode for which the AS will activate.
.TP
(6) \fIprofile\fR
The SL profile name that defines the SS7 characteristics of the SS7 link provided by this
application server.
.RE
.\"
.\"
.SH DIAGNOSTICS
.PP
An exit status of zero (0) indicates that the command was successful; one (1)
indicates that an error occurred and a diagnostic message is printed on
.BR stderr ;
two (2) indicates that the option syntax was in error and a diagnostic message
is printed on
.BR stderr .
.PP
The
.B --quiet
option suppresses the printing of diagnostic messages to
.BR stderr .
.\"
.\"
.SH NOTICES
.PP
Note that while it is possible under the M2UA specifications to have one ASP being served by
multiple SGs, the M2UA-AS driver treats this as multiple ASPs from the point of view of
configuration.  However, because ASPs are logical entities, although the driver treats each ASP-SG
relation to be a separate ASP, they can be viewed externally as the same entity.
.\"
.\"
.\".SH EXAMPLES
.\".PP
.\"
.\"
.SH DEVICES
.PP
.BR m2ua (4).
.\"
.\"
.SH FILES
.PP
.TP
.B /etc/sysconfig/sigtran/m2uaasd.conf
The
.B m2uaasd
configuration file.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR m2ua_as (4).
.\"
.\"
.SH BUGS
.PP
.B m2uaasd
has no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.B m2uaasd
is an OpenSS7-specific configuration command.  Because this command is used at startup and is
transparent to the SL-User opening M2UA-AS streams, compatibility is not an issue.
.\"
.\"
.SH CONFORMANCE
.PP
System configuration programs are not subject to standardization.
.B m2uaasd
conforms to OpenSS7 specifications.
.\"
.\"
.SH HISTORY
.PP
.B m2uaasd
first appeared in the OpenSS7
.B sigtran-0.9.2.2
package.
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/util/m2uaasd.c
/*****************************************************************************

 @(#) File: src/util/m2uaasd.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/util/m2uaasd.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 *  This is a configuration daemon for the M2UA-AS multiplexing driver.  Its purpose is to open a
 *  configuration file, establish SCTP associations between the local host and remote SG hosts, link
 *  the SCTP associations under the multiplexing driver at the appropriate ASP positions, configure
 *  each ASP using input-output controls and then wait in the background listening for management
 *  events of a management stream and also wait for updates to the configuration file.
 */

int
main(int argc, char **argv)
{
	int c;
	int val;

	while (1) {
		int option_index = 0;
		/* *INDENT-OFF* */
		static struct option long_options[] = {
			{"quiet",	no_argument,		NULL, 'q'},
			{"verbose",	optional_argument,	NULL, 'v'},
			{"help",	no_argument,		NULL, 'h'},
			{"version",	no_argument,		NULL, 'V'},
			{"copying",	no_argument,		NULL, 'C'},
			{NULL,		0,			NULL,  0 }
		};
		/* *INDENT-ON* */

		c = getopt_long(argc, argv, "qvhVC?", long_options, &option_index);
		if (c == -1)
			break;
		switch (c) {
		case 'q':	/* -q, --quiet */
			output -= output > 0 ? 1 : output;
			break;
		case 'v':	/* -v, --versbose */
			if (optarg == NULL) {
				output += 1;
				break;
			}
			if ((val = strtol(optarg, NULL, 0)) < 0)
				goto bad_option;
			output = val;
			break;
		case 'h':	/* -h, --help */
			help(argc, argv);
			exit(0);
		case 'V':	/* -V, --version */
			version(argc, argv);
			exit(0);
		case 'C':	/* -C, --copying */
			copying(argc, argv);
			exit(0);
		case '?':
		default:
		      bad_option:
			// syntax_error:
			if (optind < argc) {
				fprintf(stderr, "%s: illegal syntax -- ", argv[0]);
				for (; optind < argc; optind++)
					fprintf(stderr, "%s ", argv[optind]);
				fprintf(stderr, "\n");
			}
			usage(argc, argv);
			exit(2);
		}
	}
	exit(0);
}
